src/src/lms/lms.service.ts

```diff
@@ -419,5 +374,370 @@ export class LMSService {
+							'' +
+							user_id +
+							'/' +
+							test_id;
+
+						let modifiedHtml = null;
+						const modified = await new Promise(
+							(resolve, reject) => {
+								qr.toDataURL(
+									qr_code_verify_link,
+									function (err, code) {
+										if (err) {
+											resolve(null);
+											return;
+										}
+
+										if (code) {
+											const newHtml = code;
+
+											const root = parse(temp_html_code);
+
+											// Find the img tag with id "qrcode"
+											const qrcodeImg =
+												root.querySelector(
+													'#qr_certificate',
+												);
+
+											if (qrcodeImg) {
+												qrcodeImg.setAttribute(
+													'src',
+													newHtml,
+												);
+												modifiedHtml = root.toString();
+
+												resolve(modifiedHtml);
+											} else {
+												resolve(null);
+											}
+										} else {
+											resolve(null);
+										}
+									},
+								);
+							},
+						);
+						if (modifiedHtml != null) {
+							temp_html_code = modifiedHtml;
+						}
+					} catch (e) {
+						console.log(e);
+					}
+					const lmsCertificate = new LMSCertificateDto({
+						user_id: user_id,
+						test_id: test_id,
+						certificate_status: 'Issued',
+						issuance_date: issuance_date,
+						expiration_date: expiration_date,
+						certificate_html: temp_html_code,
+					});
+					await this.issueCertificateHtml(lmsCertificate);
+				}
+				// Update in attendance data in database
+				await this.markCertificateStatus(
+					userTestData?.id,
+					issue_status,
+				);
+			}
+		}
+	}
+	async fetchTestTrackingData(limit: number) {
+		const query = `
+			query Getlms_test_tracking {
+				lms_test_tracking(
+				where:{
+					certificate_status:{
+						_is_null: true
+					}
+				},
+				limit: ${limit}
+				){
+					id
+					user_id
+					test_id
+					score
+					context
+					context_id
+				}
+			}
+			`;
+		try {
+			const data_list = (await this.hasuraService.getData({ query }))
+				?.data?.lms_test_tracking;
+			//console.log('data_list cunt------>>>>>', data_list.length);
+			//console.log('data_list------>>>>>', data_list);
+			return data_list;
+		} catch (error) {
+			console.log('fetchTestTrackingData:', error, error.stack);
+			return [];
+		}
+	}
+	async getUserAttendanceList(user_id, context, context_id) {
+		const query = `query MyQuery {
+				attendance(where: {user_id: {_eq: ${user_id}}, context: {_eq: ${context}}, context_id: {_eq:${context_id}}, status: {_eq: "present"}}) {
+					id
+					status
+					context
+					context_id
+				  }
+			  }`;
+		try {
+			const data_list = (await this.hasuraService.getData({ query }))
+				?.data?.attendance;
+			//console.log('data_list cunt------>>>>>', data_list.length);
+			//console.log('data_list------>>>>>', data_list);
+			return data_list;
+		} catch (error) {
+			console.log('getUserAttendanceList:', error, error.stack);
+			return [];
+		}
+	}
+	async getUserName(user_id) {
```

Comment:
@rushi-tekdi Check if EG has existing function for this

Reply:
moved